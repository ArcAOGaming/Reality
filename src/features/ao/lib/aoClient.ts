import { connect, result } from "@permaweb/aoconnect";

export type MessageId = string;

export type Message = {
  Target: string;
  Data: string;
};

type BaseAoClient = ReturnType<typeof connect>;

export type AoClient = BaseAoClient;

export type MessageResult = Awaited<ReturnType<typeof result>>;

export type MessageWithResult = {
  messageId: MessageId;
  result: MessageResult;
};
