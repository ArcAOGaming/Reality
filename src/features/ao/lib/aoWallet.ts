import { createDataItemSigner } from "@permaweb/aoconnect";

export type AoSigner = ReturnType<typeof createDataItemSigner>

export type AoWallet = {
  address: string;
  signer: AoSigner;
}

export type AoWalletConnectionResult = {
  success: true;
  result: AoWallet;
} | {
  success: false;
  error: unknown;
}

// TODO: Import these from somewhere more official
export type PermissionType =
  | "ACCESS_ADDRESS"
  | "ACCESS_PUBLIC_KEY"
  | "ACCESS_ALL_ADDRESSES"
  | "SIGN_TRANSACTION"
  | "ENCRYPT"
  | "DECRYPT"
  | "SIGNATURE"
  | "ACCESS_ARWEAVE_CONFIG"
  | "DISPATCH";

export interface AppInfo {
  name?: string;
  logo?: string;
}

export interface GatewayConfig {
  host: string;
  port: number;
  protocol: "http" | "https";
}

export type AoWalletConnector = (
  permissions: PermissionType[],
  appInfo?: AppInfo,
  gateway?: GatewayConfig,
) => Promise<AoWalletConnectionResult>;
