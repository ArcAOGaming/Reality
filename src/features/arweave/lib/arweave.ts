import Arweave from "arweave";

export const defaultArweave = Arweave.init({});
