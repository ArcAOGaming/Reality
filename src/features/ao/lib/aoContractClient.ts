import { z } from "zod";
import { AoClient, Message } from "./aoClient";
import { AoWallet } from "./aoWallet";
import { connect } from "@permaweb/aoconnect";

class AoContractError extends Error {
  error: unknown | undefined;

  constructor(message: string, error?: unknown) {
    super(message);
    this.name = "AoContractError";
    this.error = error;
  }
}

export type ReadArgs = Pick<Parameters<AoClient["dryrun"]>[0], "tags" | "data" | "anchor">

export type SendArgs = Pick<Parameters<AoClient["message"]>[0], "tags" | "data" | "anchor">

export type AoContractClient = {
  processId: string;
  aoClient: AoClient;
  aoWallet: AoWallet;

  dryrunReadReplyOptional: (readArgs: ReadArgs) => Promise<Message | undefined>;
  dryrunReadReplyOne: (readArgs: ReadArgs) => Promise<Message>;
  dryrunReadReplyOneJson: <T>(readArgs: ReadArgs, schema?: z.Schema) => Promise<T>;
  message: (sendArgs: SendArgs) => Promise<string>;
}

export const createAoContractClient = (
  processId: string,
  aoClient: AoClient,
  aoWallet: AoWallet,
): AoContractClient => {
  const dryrunReadReplyOptional = async (readArgs: ReadArgs) => {
    const result = await aoClient.dryrun({
      ...readArgs,
      process: processId,
      Target: processId,
      Owner: aoWallet.address,
    });
    const messages = result.Messages as Array<Message>;

    if (messages.length === 0) {
      return undefined;
    }

    const reply = messages.find((msg) => msg.Target === aoWallet.address);
    return reply;
  }

  const dryrunReadReplyOne = async (readArgs: ReadArgs) => {
    const reply = await dryrunReadReplyOptional(readArgs);
    if (!reply) {
      throw new AoContractError("No reply");
    }
    return reply;
  }

  const dryrunReadReplyOneJson = async (readArgs: ReadArgs, schema?: z.Schema) => {
    const reply = await dryrunReadReplyOne(readArgs);
    
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let json: any;
    try {
      json = JSON.parse(reply.Data);
    } catch (error) {
      throw new AoContractError("Invalid JSON", error);
    }
    if (schema) {
      const result = schema.safeParse(json);
      if (!result.success) {
        throw new AoContractError("JSON does not match schema", result.error);
      }
      return result.data;
    }
    return json;
  }

  const message = async (sendArgs: SendArgs) => 
    aoClient.message({
      ...sendArgs,
      process: processId,
      signer: aoWallet.signer,
    });

  return {
    processId,
    aoClient,
    aoWallet,

    dryrunReadReplyOptional,
    dryrunReadReplyOne,
    dryrunReadReplyOneJson,
    message,
  }
}

export const createAoContractClientForProcess = (wallet: AoWallet) => (processId: string) => {
  const aoClient = connect();
  return createAoContractClient(processId, aoClient, wallet);
}
export type AoContractClientForProcess = ReturnType<typeof createAoContractClientForProcess>;
