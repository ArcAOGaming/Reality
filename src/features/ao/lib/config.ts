import { PermissionType } from "./aoWallet"

export const requiredPermissions: Array<PermissionType> = [
  "ACCESS_ADDRESS",
  "SIGN_TRANSACTION",
  // "ACCESS_PUBLIC_KEY",
  // "ENCRYPT",
  // "DECRYPT",
  "SIGNATURE",
  "ACCESS_ARWEAVE_CONFIG",
]

export const requestedPermissions = [
  ...requiredPermissions,
  // ...optionalPermissions,
]
