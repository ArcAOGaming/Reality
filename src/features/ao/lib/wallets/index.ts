import { AoWalletConnector } from "../aoWallet";
import { connectInjectedWallet } from "./injected";
import { connectOthentWallet } from "./othent";
import { createAnonymousWallet } from "./anonymous";

export const WalletType = ["Injected", "Othent", "Keyfile"] as const;
export type WalletType = typeof WalletType[number];

export  const wallets: Record<WalletType, AoWalletConnector> = {
  "Injected": connectInjectedWallet,
  "Othent": connectOthentWallet,
  "Keyfile": createAnonymousWallet,
} as const;
