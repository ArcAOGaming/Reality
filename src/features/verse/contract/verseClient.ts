import { MessageId } from "../../ao/lib/aoClient";
import { AoContractClient, createAoContractClient } from "../../ao/lib/aoContractClient";
import { VerseEntities, VerseEntity, VerseEntityCreate, VerseEntityPosition, VerseInfo, VerseParameters } from "./model";
import { AoWallet } from "@/features/ao/lib/aoWallet";
import { connect } from "@permaweb/aoconnect";

export type VerseClient = {
  aoContractClient: AoContractClient;

  verseId: string;

  // Reads
  readInfo(): Promise<VerseInfo>;
  readParameters(): Promise<VerseParameters>;
  readEntitiesStatic(): Promise<VerseEntities>;
  readEntitiesDynamic(sinceTime: Date): Promise<VerseEntities>;

  // Writes
  createEntity(entity: VerseEntity): Promise<MessageId>;
  updateEntityPosition(position: VerseEntityPosition): Promise<MessageId>;
}

export const createVerseClient = (
  aoContractClient: AoContractClient,
): VerseClient => ({
  aoContractClient: aoContractClient,

  verseId: aoContractClient.processId,

  // Read
  readInfo: () => aoContractClient.dryrunReadReplyOneJson<VerseInfo>({
    tags: [{ name: "Action", value: "VerseInfo" }]
  }, /* VerseInfo */),
  readParameters: () => aoContractClient.dryrunReadReplyOneJson<VerseParameters>({
    tags: [{ name: "Action", value: "VerseParameters" }]
  }, /* VerseParameters */), // TODO: Define VerseParameters properly
  readEntitiesStatic: () => aoContractClient.dryrunReadReplyOneJson<VerseEntities>({
    tags: [{ name: "Action", value: "VerseEntitiesStatic" }]
  }, /* VerseEntities */), // TODO: Define VerseEntities properly
  readEntitiesDynamic: (sinceTime: Date) => aoContractClient.dryrunReadReplyOneJson<VerseEntities>({
    tags: [{ name: "Action", value: "VerseEntitiesDynamic" }],
    // TODO: Get timestamp working!!
    data: JSON.stringify({ Timestamp: Math.floor(sinceTime.getTime()) }),
  }, /* VerseEntities */), // TODO: Define VerseEntities properly

  // Write
  createEntity: (entity: VerseEntityCreate) => aoContractClient.message({
    tags: [{ name: "Action", value: "VerseEntityCreate" }],
    data: JSON.stringify(entity),
  }),
  updateEntityPosition: (position: VerseEntityPosition) => aoContractClient.message({
    tags: [{ name: "Action", value: "VerseEntityUpdatePosition" }],
    data: JSON.stringify({ Position: position }),
  }),
});

export type VerseClientForProcess = (processId: string) => VerseClient;

export const createVerseClientForProcess = (wallet: AoWallet): VerseClientForProcess => (processId: string) => {
  const aoContractClient = createAoContractClient(processId, connect(), wallet);
  return createVerseClient(aoContractClient);
}
