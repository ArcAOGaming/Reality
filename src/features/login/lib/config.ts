import { ConnectConfig } from "@/features/ao/lib/aoWallet";
import { permissionsRequested, permissionsRequired } from "@/features/ao/lib/config";

export const defaultConnectConfig: ConnectConfig = {
  permissionsRequested,
  permissionsRequired,
  appInfo: {
    name: "WeaveWorld",
    // TODO: Add logo
    // logo: 
  },
}

export const dummyConnectConfig: ConnectConfig = {
  permissionsRequested: [],
}
