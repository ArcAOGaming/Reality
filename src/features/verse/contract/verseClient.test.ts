import { describe, test, expect, beforeAll } from "vitest";
import { createVerseClient } from "./verseClient";
import { AoContractClient, createAoContractClient } from "@/features/ao/lib/aoContractClient";
import { loadTestWallet } from "@/features/ao/test/lib/fsWallet";
import { AoWallet } from "@/features/ao/lib/aoWallet";
import { connect } from "@permaweb/aoconnect";


describe('createVerseClient', () => {
  let testWallet: AoWallet;
  let testAoContractClient: AoContractClient;
  let universeAoContractClient: AoContractClient;

  beforeAll(async () => {
    testWallet = await loadTestWallet();
    testAoContractClient = createAoContractClient(import.meta.env.VITE_READ_PROCESS_ID, connect(), testWallet);
    universeAoContractClient = createAoContractClient(import.meta.env.VITE_UNIVERSE_PROCESS_ID, connect(), testWallet);
  })

  test('creates client', async () => {
    const client = createVerseClient(testAoContractClient)
    expect(client).toMatchSnapshot();
  })

  test('Universe readInfo', async () => {
    const client = createVerseClient(universeAoContractClient)
    const info = await client.readInfo();
    expect(info).toMatchSnapshot();
  })
  
  test('Universe readParameters', async () => {
    const client = createVerseClient(universeAoContractClient)
    const info = await client.readParameters();
    expect(info).toMatchSnapshot();
  })

  test('Universe readAllEntities', async () => {
    const client = createVerseClient(universeAoContractClient)
    const info = await client.readAllEntities();
    expect(info).toMatchSnapshot();
  })
})
