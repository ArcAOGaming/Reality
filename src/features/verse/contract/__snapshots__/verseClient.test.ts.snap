// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVerseClient > Universe readAllEntities 1`] = `
{
  "BSs3O9QCfi61LihXOBQn8I3uyS_bsoFHa8cccBmfuYY": {
    "Position": [
      0,
    ],
    "Type": "Warp",
  },
}
`;

exports[`createVerseClient > Universe readInfo 1`] = `
{
  "Dimensions": 1,
  "Name": "Universe",
  "Parent": "An5_4dya7LA2PICud62nxu9k-d7UgnUlvMsU6mIdknI",
  "Render-With": "1D-Directory",
}
`;

exports[`createVerseClient > Universe readParameters 1`] = `{}`;

exports[`createVerseClient > creates client 1`] = `
{
  "aoContractClient": {
    "aoClient": {
      "assign": [Function],
      "dryrun": [Function],
      "message": [Function],
      "monitor": [Function],
      "result": [Function],
      "results": [Function],
      "spawn": [Function],
      "unmonitor": [Function],
    },
    "aoWallet": {
      "address": "HHz5nNVEzkALjmZq8cdIdubMBJCpIFOvzO5uHfU5nQI",
      "anonymous": false,
      "signer": [Function],
      "type": "Keyfile",
    },
    "dryrunReadReplyOne": [Function],
    "dryrunReadReplyOneJson": [Function],
    "dryrunReadReplyOptional": [Function],
    "message": [Function],
    "processId": "PHHOfjFPSbn7TQT1Z-dB4ndyIyppHV5haMPCg8dIccs",
  },
  "readAllEntities": [Function],
  "readEntities": [Function],
  "readInfo": [Function],
  "readParameters": [Function],
  "writePosition": [Function],
}
`;

exports[`createVerseClient > univrese readInfo 1`] = `
{
  "Dimensions": 1,
  "Name": "Universe",
  "Parent": "An5_4dya7LA2PICud62nxu9k-d7UgnUlvMsU6mIdknI",
  "Render-With": "1D-Directory",
}
`;
